moksha.hub/moksha/hub/stomp/stomp.py

```diff
@@ -155,12 +156,13 @@ def failover(self):
 
     def start_heartbeat(self, interval):
         self._heartbeat_enabled = True
-        reactor.callLater(interval / 1000.0, self.heartbeat, interval)
+        reactor.callInThread(self.heartbeat, interval)
```

Comment:
At a glance, this doesn't look right. Nothing in ``self.heartbeat`` blocks, so deferring it to a thread is not only unnecessary, it's unsafe because ``self.proto`` is (presumably) a Twisted protocol which you should *never* call outside the reactor thread. If you're finding that callLater isn't called around when you want it to, it's because something is running blocking code in the reactor thread which is when you should use this API.

Reply:
On the first point: see that a new (disappointing) `time.sleep` call is added in `self.heartbeat`.

On the second point: I have the same worry.

Reply:
Well, you can't rely on interacting with the protocol object outside the twisted thread, it's not thread-safe. You could call back in with deferFromThread (or whatever it's called), but you're going to run into the same scheduling issue you're trying to avoid. Whatever is resulting in blocking calls (the consumers?) needs to run it its own thread.